src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -0,0 +1,166 @@
+            ProjectSnapshotManagerBase projectSnapshotManager)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (projectSnapshotManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManager));
+            }
+
+            _dotNetProject = project;
+            _foregroundDispatcher = foregroundDispatcher;
+            _projectSnapshotManager = projectSnapshotManager;
+            _projectChangedSemaphore = new SemaphoreSlim(initialCount: 1);
+
+            AttachToProject();
+        }
+
+        public DotNetProject DotNetProject => _dotNetProject;
+
+        public HostProject HostProject => _currentHostProject;
+
+        protected ForegroundDispatcher ForegroundDispatcher => _foregroundDispatcher;
+
+        public void Detatch()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified -= DotNetProject_Modified;
+
+            UpdateHostProjectForeground(null);
+        }
+
+        protected abstract Task OnProjectChangedAsync();
+
+        // Protected virtual for testing
+        protected virtual void AttachToProject()
+        {
+            ForegroundDispatcher.AssertForegroundThread();
+
+            DotNetProject.Modified += DotNetProject_Modified;
+
+            // Trigger the initial update to the project.
+            _batchingProjectChanges = true;
+            Task.Factory.StartNew(ProjectChangedBackgroundAsync, null, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.BackgroundScheduler);
+        }
+
+        // Must be called inside the lock.
+        protected async Task UpdateHostProjectUnsafeAsync(HostProject newHostProject)
+        {
+            if (_foregroundDispatcher.IsForegroundThread)
+            {
+                UpdateHostProjectForeground(newHostProject);
+            }
+            else
+            {
+                await Task.Factory.StartNew(UpdateHostProjectForeground, newHostProject, CancellationToken.None, TaskCreationOptions.None, ForegroundDispatcher.ForegroundScheduler);
+            }
+        }
+
+        protected async Task ExecuteWithLockAsync(Func<Task> func)
+        {
+            await _projectChangedSemaphore.WaitAsync().ConfigureAwait(false);
+
+            await func().ConfigureAwait(false);
+
+            _projectChangedSemaphore.Release();
+        }
+
+        private async Task ProjectChangedBackgroundAsync(object state)
+        {
+            ForegroundDispatcher.AssertBackgroundThread();
+
+            _batchingProjectChanges = false;
+
+            await OnProjectChangedAsync();
+        }
+
+        private void DotNetProject_Modified(object sender, SolutionItemModifiedEventArgs args)
```

Comment:
Do they have a batching construct or do they just call this a million times?

Reply:
They call a bajillion times